src/librustc_session/feature_gate.rs

```diff
@@ -0,0 +1,94 @@
+use syntax_pos::edition::Edition;
```

Comment:
My reaction to this is that I would prefer the following layout for feature gating:

- A crate dedicated to declaring feature gates and the structs ("data crate"); this would consist of `active.rs`, `accepted.rs`, `removed.rs`, most of `builtin_attrs.rs` (at least the parts not dependent on `libsyntax`), the structs from `mod.rs`). Let's call this crate `librustc_feature` and we'd have `librustc_session` depend on it.

- A module for checking (`check.rs` -> `gate_features.rs`), possibly as part of `librustc_passes`.

Reply:
Implemented the first part of this in https://github.com/rust-lang/rust/pull/66895, which should be sufficient for this PR to build upon.

Reply:
Will rebase atop that.